File: golang-backend/internal/handlers/projecthandler/dto.go
package projecthandler

import (
	"github.com/aportela/doneo/internal/handlers"
	"github.com/aportela/doneo/internal/handlers/projectpriorityhandler"
	"github.com/aportela/doneo/internal/handlers/projectstatushandler"
	"github.com/aportela/doneo/internal/handlers/projecttypehandler"
	"github.com/aportela/doneo/internal/handlers/userhandler"
)

type projectType struct {
	ID string `json:"id"`
}

type projectPriority struct {
	ID string `json:"id"`
}

type projectStatus struct {
	ID string `json:"id"`
}

type addRequest struct {
	ID          string          `json:"id"`
	Key         string          `json:"key"`
	Summary     string          `json:"summary"`
	Description *string         `json:"description"`
	Type        projectType     `json:"type"`
	Priority    projectPriority `json:"priority"`
	Status      projectStatus   `json:"status"`
}

type updateRequest struct {
	ID          string          `json:"id"`
	Key         string          `json:"key"`
	Summary     string          `json:"summary"`
	Description *string         `json:"description"`
	Type        projectType     `json:"type"`
	Priority    projectPriority `json:"priority"`
	Status      projectStatus   `json:"status"`
}

type filterRequest struct {
	Key *string `json:"key"`
}

type searchRequest struct {
	Pager  handlers.PagerRequest `json:"pager"`
	Order  handlers.OrderRequest `json:"order"`
	Filter *filterRequest        `json:"filter"`
}

type projectResponse struct {
	ID                     string                                         `json:"id"`
	Key                    string                                         `json:"key"`
	Summary                string                                         `json:"summary"`
	Description            string                                         `json:"description"`
	CreatedBy              userhandler.UserBaseResponse                   `json:"createdBy"`
	CreatedAt              int64                                          `json:"createdAt"`
	Type                   projecttypehandler.ProjectTypeResponse         `json:"type"`
	Priority               projectpriorityhandler.ProjectPriorityResponse `json:"priority"`
	Status                 projectstatushandler.ProjectStatusResponse     `json:"status"`
	TasksCount             uint                                           `json:"tasksCount"`
	PermissionsCount       uint                                           `json:"permissionsCount"`
	AttachmentsCount       uint                                           `json:"attachmentsCount"`
	NotesCount             uint                                           `json:"notesCount"`
	HistoryOperationsCount uint                                           `json:"historyOperationsCount"`
}

type searchProjectsResponse struct {
	Projects []projectResponse      `json:"projects"`
	Pager    handlers.PagerResponse `json:"pager"`
}

File: vite-vue-ts-frontend/src/modules/projects/types/dto.ts
import type {
  PagerRequest,
  PagerResponse,
} from "../../../shared/types/dto/pager";
import type { Order } from "../../../shared/types/dto/order";

import type { ProjectTypeResponse } from "../../project-types/types/dto";
import type { ProjectPriorityResponse } from "../../project-priorities/types/dto";
import type { ProjectStatusResponse } from "../../project-statuses/types/dto";
import type { UserBaseResponse } from "../../users/types/dto";

export type AddRequest = {
  key: string;
  summary: string;
  description: string | null;
  type: {
    id: string;
  };
  priority: {
    id: string;
  };
  status: {
    id: string;
  };
};

export type UpdateRequest = {
  id: string;
  key: string;
  summary: string;
  description: string | null;
  type: {
    id: string;
  };
  priority: {
    id: string;
  };
  status: {
    id: string;
  };
};

type SearchRequestFilter = {
  key?: string;
};

export type SearchRequest = {
  pager: PagerRequest;
  order: Order;
  filter?: SearchRequestFilter;
};

export type ProjectResponse = {
  id: string;
  key: string;
  summary: string;
  description: string;
  type: ProjectTypeResponse;
  priority: ProjectPriorityResponse;
  status: ProjectStatusResponse;
  createdAt: number;
  createdBy: UserBaseResponse;
  tasksCount: number;
  permissionsCount: number;
  attachmentsCount: number;
  notesCount: number;
  historyOperationsCount: number;
};

export type SearchResponse = {
  projects: ProjectResponse[];
  pager: PagerResponse;
};
